### PR TITLE
Package manager is not allowed to override std lib packages

### PR DESCRIPTION
**Pull request by pinzart90**
### Purpose

https://jira.autodesk.com/browse/DYN-3459

Changes:
**Bug Fix:**
PackageVersionHeader responses now have the `name` attribute and we can compare against any existing local package (from disk).

**New functionality:**
When installing a new package from the Package Manager client, users will get a error message box about conflicts with packages from standard library. 

1. If a user tries to install a package of the same exact version - that is already installed in the std.lib - they will get a warning saying that the package already exists as part of the std lib and the desired package will not be installed.

2. If a user tries to install a different version of a package that is already installed in the std.lib - they will get a warning saying they cannot override it and the desired version will not be installed

3. If a user tries to install a package as part of a dep chain from another package - that is already installed in the std.lib - they will get a warning saying they cannot override it (because one of the deps is a part of std lib) and the desired package will not be installed

The user will not be allowed to install packages that are in conflict with the std lib. This is temporary until we fix the following taks:
Further UI enhancements will be attempted in another jira task https://jira.autodesk.com/browse/DYN-3600

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:534]: this is the change that fixes part of the duplicate packages problem.
It was broken with this change https://github.com/DynamoDS/Dynamo/pull/10544/files#diff-10009027dea2b20f51c89c7b0306d25ac8d92dbfc0fbfb891fdfa171eba8df1fR707

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:593]: I moved this so that the std lib conflict err message can appear before any others (like python scripts etc)

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:739]: same issue with ID vs Name
broken here https://github.com/DynamoDS/Dynamo/pull/10544/files#diff-10009027dea2b20f51c89c7b0306d25ac8d92dbfc0fbfb891fdfa171eba8df1fR707

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (mjkkirschner) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:534]: so there is this note in packgeManagerClientViewModel that names will sometimes be null:
https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs#L654

will that now be a problem with this change in comparison?

Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:534]: @mmisol can you explain how that can happen ?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:739]: Can you give an example of what a package `id` looks like vs. a package `name`?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:535]: typo: `duplicateLocalPackages`

**Comment by aparajit-pratap** [comment]
@pinzart90 is it possible to write tests for this?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:514]: this change need PMclient and PMserver changes...Waiting for approval

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:567]: the `if (stdLibPackages.Any())` block has the only functionality changes I added.
The for loop `foreach (var localPkg in duplicateLoacalPackages)` and all the Lists of packages (            `uninstallsRequiringRestart`, `uninstallRequiringUserModifications`, `immediateUninstalls` ) were already heder. I just moved them. I am not sure how relevant they still are .... but I do not think they conflict with the std lib logic.

As far as I can tell...`immediateUninstalls` should be localPackages that are not loaded or in use...so they can be uninstalled (deleted from disk) without restart.
`uninstallRequiringUserModifications` looks to be the case when the user has workspaces open (from a particular package)...

**Comment by mjkkirschner** [comment]
Immediate uninstalls can also be custom node only packages (.dyf) since there is no binary to mess up unloading

**Comment by pinzart90** [comment]
PM server changes (https://git.autodesk.com/Dynamo/PackageManager/pull/307) and PM client changes (https://git.autodesk.com/Dynamo/PackageManagerClient/commit/2cebc8f0b95f891ac7740816b190f10e9eb721b2) have been merged
@mjkkirschner PTAL

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:747]: the ID is the PM server's ID (a GUID)
The name is any package's name

File: test/DynamoCoreWpfTests/PackageManagerUITests.cs
Comment (mjkkirschner) [inline test/DynamoCoreWpfTests/PackageManagerUITests.cs:158]: if you want to stick with a general name can we add a comment describing the test a bit?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs]: The package itself is included in the `full_dependency_ids` array

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:552]: explained here https://github.com/DynamoDS/Dynamo/pull/11580/#discussion_r604214702

**Comment by pinzart90** [comment]
@mjkkirschner @aparajit-pratap I updated the PR description to include what changed.

File: src/DynamoCoreWpf/Properties/Resources.en-US.resx
Comment (pinzart90) [inline src/DynamoCoreWpf/Properties/Resources.en-US.resx:2356]: {0}  = "Dynamo"
{1} = The desired package name, ex: MeshToolkit
{2} = List of package names and versions are in conflict.

**Review by mjkkirschner — APPROVED** [review]

**Comment by mjkkirschner** [comment]
this passed here:
https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/2360/parameters/
not sure whats up with the check not updating.
@pinzart90 

@aparajit-pratap @sm6srw what do you think about merging this?

File: src/DynamoCoreWpf/Properties/Resources.en-US.resx
Comment (aparajit-pratap) [reply src/DynamoCoreWpf/Properties/Resources.en-US.resx:2356]: Does 0 come from branding resource provider or something like that?

File: src/DynamoCoreWpf/Utilities/MessageBoxUtilities.cs
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/Utilities/MessageBoxUtilities.cs:26]: Why can't we use the `moq` library for this in the tests, why do we need to implement our own classes to mock the message box?

Comment (pinzart90) [reply src/DynamoCoreWpf/Utilities/MessageBoxUtilities.cs:26]: `moq` only works on interfaces. The only way to moq the MessageBox class is to wrap it in an interface.

File: src/DynamoCoreWpf/Properties/Resources.en-US.resx
Comment (pinzart90) [reply src/DynamoCoreWpf/Properties/Resources.en-US.resx:2356]: string at position 0 is the application name( Ex. Dynamo, Dynamo for Revit)

Comment (aparajit-pratap) [reply src/DynamoCoreWpf/Properties/Resources.en-US.resx:2356]: Yes, I know that. Never mind my question.

**Review by sm6srw — APPROVED** [review]
LGTM!

**Comment by mjkkirschner** [comment]
merging!

**Pull request merged**
